src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+    return stencilCallsInOrder_;
+  }
+
+  void visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) {
+    stencilCallsInOrder_.push_back(stmt);
+  }
+};
+
+/// @brief The AddBoundaryConditions class traverses the StencilDescAST to extract all the
+/// StencilCallStmts for a stencili with a given ID. This is required to properly insert boundary
+/// conditions.
+class AddBoundaryConditions : public ASTVisitorForwarding {
+  StencilInstantiation* instantiation_;
+  int StencilID_;
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>> stencilCallsToReplace_;
+
+public:
+  AddBoundaryConditions(StencilInstantiation* instantiation, int StencilID)
+      : instantiation_(instantiation), StencilID_(StencilID) {}
+
+  void visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) override {
+    if(instantiation_->getStencilIDFromStmt(stmt) == StencilID_)
+      stencilCallsToReplace_.emplace_back(stmt);
+  }
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCallsToReplace() {
+    return stencilCallsToReplace_;
+  }
+
+  void reset() { stencilCallsToReplace_.clear(); }
+};
+
+PassSetBoundaryCondition::PassSetBoundaryCondition() : Pass("PassSetBoundaryCondition") {}
+
+bool PassSetBoundaryCondition::run(StencilInstantiation* stencilInstantiation) {
+
+  // check if we need to run this pass
+  if(stencilInstantiation->getStencils().size() == 1) {
+    if(stencilInstantiation->getOptimizerContext()->getOptions().ReportBoundaryConditions) {
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName() << " :";
+      std::cout << " No boundary conditions applied\n";
+    }
+    return true;
+  }
+  // returns the original ID of a variable
+  auto getOriginalID = [&](int ID) {
+    // check if the variable exists to be sure not to assert
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(
+        stencilInstantiation->getOriginalNameFromAccessID(ID));
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      if(stencilInstantiation->isField(ID)) {
+        return stencilInstantiation->getAccessIDFromName(
+            stencilInstantiation->getOriginalNameFromAccessID(ID));
+      } else {
+        return (int) FieldType::FT_NotOriginal;
+      }
+    } else {
+      return (int) FieldType::FT_NotOriginal;
+    }
+
+  };
+
+  std::unordered_map<int, Extents> dirtyFields;
+  std::unordered_map<int, std::shared_ptr<BoundaryConditionDeclStmt>> allBCs;
+
+  // Fetch all the boundary conditions stored in the instantiation
+  for(const auto& bc : stencilInstantiation->getBoundaryConditions()) {
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(bc.first);
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      int ID = stencilInstantiation->getAccessIDFromName(bc.first);
+      allBCs.emplace(ID, bc.second);
+    }
+  }
+
+  // Get the order in which the stencils are called:
+  VisitStencilCalls findStencilCalls;
```

Comment:
Here you want to obtain a list of all the stencils executed. Are you sure what you want to do is to visit the StencilCalls? 
Did you think what happen if the stencil call is inlined or fused? 
Isnt `StencilInstantiation::stencils_` the list that you need?

Reply:
can we be certain that the order (and potential repetitions) there is the right one? By fetching it form the StencilDescAST, it is definilty the right one. I am pretty sure this is the right way to go